Fix Paginating, Sorting, and Searching Issues Within  "Research Outputs" Tab

aaronskiba opened the pull request:
Fixes #935

Changes proposed in this PR:
- Add `scope :search` to `ResearchOutput` model: e289fba9b0cbe4af4699455cf08adca8f597dac1
  - Search query is performed by `research_outputs.title` (case-insensitive) 
- Create `app/controllers/paginable/research_outputs_controller.rb` 3c5a478feacc9bb9a8e69abd93579c70af29a207
  - This new controller fixes the following features on the `/plans/:plan_id/research_outputs` page:
    - The clickable pagination options (e.g. View all,1,2,Next,Last)
    - The search box (for `research_outputs.title` matches)
    - The clickable sorting arrows

aaronskiba committed:
Add `scope :search` to ResearchOutput model

Search by `research_outputs.title` (case-insensitive)

aaronskiba committed:
Create `paginable/research_outputs_controller.rb`

This commit, along with e289fba9b0cbe4af4699455cf08adca8f597dac1 are being made to resolve https://github.com/portagenetwork/roadmap/issues/935.

aaronskiba committed:
Update CHANGELOG.md

lagoan commented on app/models/research_output.rb:
```diff
@@ -70,6 +70,15 @@ class ResearchOutput < ApplicationRecord
+  # = Scopes =
+  # ==========
+
+  scope :search, lambda { |term|
+                   search_pattern = "%#{term}%"
+                   where('lower(title) LIKE lower(?)', search_pattern)
+                 }
```
Are there any other values we would like to search for? Would ``abbreviation`` make sense as a search parameter?

aaronskiba replied:
![Screenshot from 2024-11-12 10-01-57](https://github.com/user-attachments/assets/b31d74f0-0ce2-41cb-8707-08bd363d492d)
I guess `type` or `repository` are possible. Although, those two aren't columns in the `research_outputs` table. At the same time, I don't imagine too many plans will have to be searched thoroughly thoroughly through this research outputs section.

lagoan commented on app/controllers/paginable/research_outputs_controller.rb:
```diff
@@ -0,0 +1,25 @@
+  class ResearchOutputsController < ApplicationController
+    include Paginable
+
+    after_action :verify_authorized
```
I like how you are protecting the controller here. I think other paginable controllers could benefit from this.

lagoan commented on app/controllers/paginable/research_outputs_controller.rb:
```diff
@@ -0,0 +1,25 @@
+      paginable_renderise(
+        partial: 'index',
+        scope: @research_outputs,
+        query_params: { sort_field: 'research_outputs.title' },
```
Do we need to pass the sort direction?

aaronskiba replied:
I can see it's being specified in other `app/controllers/paginable/` files. However, the `paginable_renderise` method we're calling seems to be sorting it in ascending order by default. (`app/controllers/concerns/paginable.rbapp/controllers/concerns/paginable.rb` includes a `require 'sort_direction'` statement and `class SortDirection` includes the following:
  ```ruby
  class SortDirection
  ##
  # When given an unknown or nil direction, default to this value
  DEFAULT_DIRECTION = 'ASC'
  ```
)

lagoan replied:
Thanks for clarifying this

lagoan commented on app/controllers/paginable/research_outputs_controller.rb:
```diff
@@ -0,0 +1,25 @@
+      # Same @research_outputs assignment as app/controllers/research_outputs_controller.rb
+      @research_outputs = ResearchOutput.includes(:repositories).where(plan_id: @plan.id)
+      # Same authorize handling as app/controllers/research_outputs_controller.rb
+      authorize @research_outputs.first || ResearchOutput.new(plan_id: @plan.id)
```
I see this authorize handling on ``app/controllers/research_outputs_controller.rb``. Why are we creating a new ``ResearchOutput`` if we cannot find one ?

aaronskiba replied:
In the case of `app/controllers/research_outputs_controller.rb`:
If I only use `authorize @research_outputs.first`, I encounter the following if the plan has no research_outputs:
```
Pundit::NotDefinedError - unable to find policy `NilClassPolicy` for `nil`:
  app/controllers/research_outputs_controller.rb:17:in `index'
```
If I use `authorize @research_outputs.first if @research_outputs.first.present?`, I encounter the following if the plan has no research_outputs:
```
Pundit::AuthorizationNotPerformedError at /plans/14224/research_outputs
ResearchOutputsController
```

However, it may be the case that there is always at least one research output when `app/controllers/paginable/research_outputs_controller.rb` is called. But it might still be safest to keep `authorize @research_outputs.first || ResearchOutput.new(plan_id: @plan.id)`? (I could at least add a comment explaining this a bit).

lagoan replied:
A comment would be good. Also add the method call parenthesis to clarify the expected order.

aaronskiba replied:
> However, it may be the case that there is always at least one research output when `app/controllers/paginable/research_outputs_controller.rb` is called. But it might still be safest to keep `authorize @research_outputs.first || ResearchOutput.new(plan_id: @plan.id)`? (I could at least add a comment explaining this a bit).

I guess `|| ResearchOutput.new(plan_id: @plan.id)` prevents `Pundit::NotDefinedError` when a direct GET /paginable/plans/:id/research_outputs request is made on a plan with 0 research_outputs.

aaronskiba replied:
> A comment would be good. Also add the method call parenthesis to clarify the expected order.

Cool, I made one more commit addressing this.

lagoan commented on app/controllers/paginable/research_outputs_controller.rb:
```diff
@@ -0,0 +1,25 @@
+
+    # GET /paginable/plans/:plan_id/research_outputs
+    def index
+      @plan = Plan.find_by(id: params[:plan_id])
+      # Same @research_outputs assignment as app/controllers/research_outputs_controller.rb
+      @research_outputs = ResearchOutput.includes(:repositories).where(plan_id: @plan.id)
```
Do we need ``@plan`` and ``@research_outputs`` to be instance variables?

aaronskiba replied:
There is a corresponding `app/views/paginable/research_outputs/_index.html.erb`. That file has usage of `@plan`, but I think `@research_outputs` can be changed to ` research_outputs`.

lagoan commented on app/controllers/paginable/research_outputs_controller.rb:
```diff
@@ -0,0 +1,25 @@
+
+    # GET /paginable/plans/:plan_id/research_outputs
+    def index
+      @plan = Plan.find_by(id: params[:plan_id])
```
Do we need to authorize by plan ? Something like https://github.com/portagenetwork/roadmap/blob/60f89e64178657c489ce40b9cb47e2c6b43a512a/app/controllers/paginable/contributors_controller.rb#L14-L15

We would need to add an application policy on ``app/policies/paginable/``

aaronskiba replied:
`app/policies/research_output_policy.rb` might be taking care of this. Here's a bit of that file's contents:
```ruby
  def index?
    @research_output.plan.readable_by?(@user.id)
  end

  def new?
    @research_output.plan.administerable_by?(@user.id)
  end

  def edit?
    @research_output.plan.administerable_by?(@user.id)
  end

  def create?
    @research_output.plan.administerable_by?(@user.id)
  end
```

lagoan reviewed (commented):
I just have some questions here to start a conversation. Sorry if I missed something.

aaronskiba committed:
Change`research_outputs` to local variable

The corresponding view, `app/views/paginable/research_outputs/_index.html.erb`, uses `scope` rather than `@research_outputs`.

aaronskiba committed:
Refactor / add comment

lagoan reviewed (approved):
![Good work!](https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExNHprbnYzc3EzZGt1Z2x1ZmVudnFyd2EycHd5N2F4cnE0aHk2c3h1ZSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/T9SzftdKiHZTzmfluU/giphy.gif)
